osquery/filesystem/fileops.h

```diff
@@ -0,0 +1,211 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#pragma once
+
+#include <fcntl.h>
+#include <string>
+#include <vector>
+
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#ifdef WIN32
+#define NOMINMAX
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+#else
+#include <unistd.h>
+#endif
+
+#include <boost/optional.hpp>
+
+namespace osquery {
+
+#ifdef WIN32
+using mode_t = int;
+using ssize_t = SSIZE_T;
+using PlatformHandle = HANDLE;
+using PlatformTimeType = FILETIME;
+
+/// Windows does not define these, X_OK on Windows just ensures that the
+/// file is readable.
+#define F_OK 0
+#define R_OK 4
+#define W_OK 2
+#define X_OK R_OK
+
+const std::string kFallbackHomeDirectory = "\\ProgramData\\osquery";
+#else
+using PlatformHandle = int;
+using PlatformTimeType = struct timeval;
+
+const std::string kFallbackHomeDirectory = "/tmp/osquery";
+#endif
+
+using PlatformTime = struct {
+  PlatformTimeType times[2];
+};
+
+/// Constant for an invalid handle
+const PlatformHandle kInvalidHandle = (PlatformHandle)-1;
+
+/**
+ * @brief File access modes for PlatformFile
+ *
+ * A file can be opened for many access modes with a variety of different
+ * options on Windows and POSIX. To provide multi-platform support, we need to
+ * provide an abstraction that can cover the supported platforms.
+ */
+
+#define PF_READ 0x0001
+#define PF_WRITE 0x0002
+
+#define PF_OPTIONS_MASK 0x001c
+#define PF_GET_OPTIONS(x) ((x & PF_OPTIONS_MASK) >> 2)
+#define PF_CREATE_NEW (0 << 2)
+#define PF_CREATE_ALWAYS (1 << 2)
+#define PF_OPEN_EXISTING (2 << 2)
+#define PF_TRUNCATE (3 << 2)
+
+#define PF_NONBLOCK 0x0020
+
+/**
+ * @brief Modes for seeking through a file
+ *
+ * Provides a chance 
+ */
+
+enum SeekMode {
+  PF_SEEK_BEGIN = 0,
+  PF_SEEK_CURRENT,
+  PF_SEEK_END
+};
+
+#ifdef WIN32
+
+/**
+ * @brief Windows-only class that deals with simulating POSIX 
+ *        asynchronous IO semantics using Windows API calls
+ */
+struct AsyncEvent {
+  AsyncEvent();
+  ~AsyncEvent();
+ 
+  OVERLAPPED overlapped_{ 0 };
```

Comment:
Is a ` ` between the `{` and end `}` in uniform initialization a thing? `clang-format` yielded this?